### ISSUE TITLE
Does not compile with g++8

### ISSUE THREAD

**Issue by lukasm91** [eth-cscs/serialbox2#157]
Compilation fails with the newest g++ version.

```
[  4%] Building CXX object src/serialbox/core/CMakeFiles/SerialboxCoreLibraryObjects.dir/FieldMap.cpp.o
cd /home/lukas/work/serialbox2/build/src/serialbox/core && /usr/bin/g++  -DSERIALBOX_ASYNC_API -DSERIALBOX_USE_EXPERIMENTAL_FILESYSTEM -isystem /home/lukas/packages/boost/1.67/include -I/home/lukas/work/serialbox2/build/src -I/home/lukas/work/serialbox2/src  -march=native -Wno-sign-compare -fPIC -O3 -DNDEBUG   -std=gnu++11 -o CMakeFiles/SerialboxCoreLibraryObjects.dir/FieldMap.cpp.o -c /home/lukas/work/serialbox2/src/serialbox/core/FieldMap.cpp
In file included from /home/lukas/work/serialbox2/src/serialbox/core/MetainfoMapImpl.h:20,
                 from /home/lukas/work/serialbox2/src/serialbox/core/FieldMetainfoImpl.h:20,
                 from /home/lukas/work/serialbox2/src/serialbox/core/FieldMap.h:18,
                 from /home/lukas/work/serialbox2/src/serialbox/core/FieldMap.cpp:15:
/home/lukas/work/serialbox2/src/serialbox/core/MetainfoValueImpl.h: In instantiation of ‘T serialbox::MetainfoValueImpl::as() const [with T = long unsigned int]’:
/home/lukas/work/serialbox2/src/serialbox/core/MetainfoValueImpl.h:87:18:   required from ‘serialbox::MetainfoValueImpl::operator T() const [with T = long unsigned int]’
/usr/include/c++/8.1.1/type_traits:932:12:   required from ‘struct std::is_constructible<long unsigned int, serialbox::MetainfoValueImpl>’
/home/lukas/work/serialbox2/src/serialbox/core/external/json/json.hpp:1476:96:   required by substitution of ‘template<class CompatibleNumberUnsignedType, typename std::enable_if<((std::is_constructible<long unsigned int, CompatibleNumberUnsignedType>::value && std::numeric_limits<_Tp>::is_integer) && (! std::numeric_limits<_Tp>::is_signed)), CompatibleNumberUnsignedType>::type <anonymous> > nlohmann::basic_json<>::basic_json(CompatibleNumberUnsignedType) [with CompatibleNumberUnsignedType = serialbox::MetainfoValueImpl; typename std::enable_if<((std::is_constructible<long unsigned int, CompatibleNumberUnsignedType>::value && std::numeric_limits<_Tp>::is_integer) && (! std::numeric_limits<_Tp>::is_signed)), CompatibleNumberUnsignedType>::type <anonymous> = <missing>]’
/usr/include/c++/8.1.1/type_traits:932:12:   required from ‘struct std::is_constructible<nlohmann::basic_json<>, serialbox::MetainfoValueImpl>’
/home/lukas/work/serialbox2/src/serialbox/core/external/json/json.hpp:1145:117:   required by substitution of ‘template<class CompatibleObjectType, typename std::enable_if<(std::is_constructible<std::__cxx11::basic_string<char>, typename CompatibleObjectType::key_type>::value && std::is_constructible<nlohmann::basic_json<>, typename CompatibleObjectType::mapped_type>::value), int>::type <anonymous> > nlohmann::basic_json<>::basic_json(const CompatibleObjectType&) [with CompatibleObjectType = serialbox::MetainfoMapImpl; typename std::enable_if<(std::is_constructible<std::__cxx11::basic_string<char>, typename CompatibleObjectType::key_type>::value && std::is_constructible<nlohmann::basic_json<>, typename CompatibleObjectType::mapped_type>::value), int>::type <anonymous> = <missing>]’
/usr/include/c++/8.1.1/ext/new_allocator.h:136:4:   [ skipping 3 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/usr/include/c++/8.1.1/bits/shared_ptr_base.h:662:4:   required from ‘std::__shared_count<_Lp>::__shared_count(std::_Sp_make_shared_tag, _Tp*, const _Alloc&, _Args&& ...) [with _Tp = serialbox::MetainfoMapImpl; _Alloc = std::allocator<serialbox::MetainfoMapImpl>; _Args = {const serialbox::MetainfoMapImpl&}; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2]’
/usr/include/c++/8.1.1/bits/shared_ptr_base.h:1328:35:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<serialbox::MetainfoMapImpl>; _Args = {const serialbox::MetainfoMapImpl&}; _Tp = serialbox::MetainfoMapImpl; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2]’
/usr/include/c++/8.1.1/bits/shared_ptr.h:360:64:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<serialbox::MetainfoMapImpl>; _Args = {const serialbox::MetainfoMapImpl&}; _Tp = serialbox::MetainfoMapImpl]’
/usr/include/c++/8.1.1/bits/shared_ptr.h:706:14:   required from ‘std::shared_ptr<_Tp> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = serialbox::MetainfoMapImpl; _Alloc = std::allocator<serialbox::MetainfoMapImpl>; _Args = {const serialbox::MetainfoMapImpl&}]’
/usr/include/c++/8.1.1/bits/shared_ptr.h:722:39:   required from ‘std::shared_ptr<_Tp> std::make_shared(_Args&& ...) [with _Tp = serialbox::MetainfoMapImpl; _Args = {const serialbox::MetainfoMapImpl&}]’
/home/lukas/work/serialbox2/src/serialbox/core/FieldMetainfoImpl.h:37:87:   required from here
/home/lukas/work/serialbox2/src/serialbox/core/MetainfoValueImpl.h:77:19: error: static assertion failed: ValueType is not supported
     static_assert(IsSupported<typename MakePrimitive<T>::type>::value,
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[1]: *** [src/serialbox/core/CMakeFiles/SerialboxCoreLibraryObjects.dir/build.make:63: src/serialbox/core/CMakeFiles/SerialboxCoreLibraryObjects.dir/FieldMap.cpp.o] Error 1
make[1]: Leaving directory '/home/lukas/work/serialbox2/build'
make: *** [Makefile:237: FieldMap.cpp.o] Error 2
```

I did not spent much time investigating this error, but I tried to track down why only the newest version of gcc shows this error; and I am not sure if this is a compiler bug or a real bug (see https://godbolt.org/g/b87VmZ).

**Comment by lukasm91** [comment]
It looks like the error is in json.hpp. We should probably update to a more recent version. It must be at least a 2.1 version (2.0.10 did not compile, 2.1.1 compiles). I did not check if there are any breaking changes, but at least for the gridtools part anything runs smoothly with this new json.hpp-version.

**Comment by rouson** [comment]
@lukasm91 I encountered the same error with GCC8.1.0.  Could you please recommend how to obtain the newer `json.hpp` and how to install it?

**Comment by lukasm91** [comment]
Sure. I simply replaced the file src/serialbox/core/external/json/json.hpp (Version 2.0.3) with https://raw.githubusercontent.com/nlohmann/json/v2.1.1/src/json.hpp (Version 2.1.1) and then compile and install as usual. I did not have any problems with that.

**Comment by rouson** [comment]
That works.  Thanks!

**Comment by havogt** [comment]
Fixed in #171
